### PR TITLE
Major Local Node Refactor

### LINKED ISSUE #418: Refactor
Currently Local Node is quite a mess, it's hard to maintain, add new features. In the past it was responsible only for starting couple of docker containers and that's it. Now apart from that, it needs to do this in different kind of modes, while activate/deactivate features, as well as generating accounts and more.

As we move on to more widely used by the community local node, we need to make sure that we provide easy to use, easy to maintain and debug project.

### PR DESCRIPTION
**Pull request by georgi-l95**
**Description**:

Currently Local Node is quite a mess, it's hard to maintain, add new features. In the past it was responsible only for starting couple of docker containers and that's it. Now apart from that, it needs to do this in different kind of modes, while activate/deactivate features, as well as generating accounts and more.

This PR aims to refactor the whole architecture by making the code easy to understand, maintain and extend.

By utilizing state machine pattern, we introduce most importantly separation of concerns, which makes sure that each and every state does only what it's supposed to do and only goes ahead to the next if it's job is completed.
And service locator pattern, allows us to easly create services which are always available due to the nature of the service locator(being singleton) and easy to reach. That way we don't have to increase coupling or duplicate code to use different kind of services.

Quick explanation:
For each configuration(start,stop,restart,generate-accounts) there is a predefined set of states (`StateData`), which are being loaded in the `StateController` upon startup. By cycling through the states, each and every one do only what it's supposed to and emit event when it's ready (event can be either finish for success or error). Most of the logic for each state is copied from the old code, it's just seprated logically. 

`Bootstraper` is used to create the Service Locator and register all services. After this all we need to do, to access a service is to call it by using the Service Locator singleton.
Current Services:

- `CLIService` is used to load all cli options and make them easy to access, whethever we need them
- `ClientService` is used to create a SDK client, which is accessable from anywhere.
- `LoggerService` is used to log all msg to the console, supports verbose and will be used together with the TUI
- `ConnectionService` have helper methods which are responsible for connection checks
- `DockerService` have all kind of docker related utils

Things that work on prevaious verions and we need to fix before release:
- TUI
- Fall back if output from docker shell execution is 1
- Readme.md update

Things that does not work on previous version and we can delay fixing, without delaying release:
- Multi node
- Debug

To start the node we can use hedera keyword (when it's released) or npm run `start`/`stop`/`restart`/`generate-accounts`
If we want to add options together with our start command we need to first add `--`. 
For example, if we want to start the node in full mode, we'd write `npm run start -- --full`

**Related issue(s)**:

Fixes #418 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by isavov — COMMENTED** [review]
Very impressive effort! 🚀

File: src/services/IService.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+export interface IService {
+
```
Comment (konstantinabl) [inline src/services/IService.ts:22]: Is this interface needed?

File: src/services/ServiceLocator.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ */
+
+import { IService } from './IService';
+
+export class ServiceLocator {
+    private services: Map<string,IService> = new Map<string, IService>();
+
+    public static Current: ServiceLocator;
+
+    public static Initiailze(): void {
```
Comment (konstantinabl) [inline src/services/ServiceLocator.ts:28]: Not sure but I think if this need to be singleton the method should be private?

File: src/services/IService.ts (hunk 1 above)
Comment (georgi-l95) [reply src/services/IService.ts:22]: Unfortunately yes. All Service are implementing this interface and when we register the services we use this interface to make sure that this is indeed a service that we want to register.

File: src/services/ServiceLocator.ts (hunk 2 above)
Comment (georgi-l95) [reply src/services/ServiceLocator.ts:28]: We use it in the Bootstrapper to initiliaze the Service Locator the first time.

**Review by konstantinabl — APPROVED** [review]

**Pull request merged**